aptos-move/block-executor/src/code_cache.rs

```diff
@@ -0,0 +1,656 @@
+                match &self.latest_view {
+                    ViewState::Sync(state) => {
+                        if let CacheRead::Hit(result) = state
+                            .captured_reads
+                            .borrow()
+                            .fetch_verified_module(address, module_name)
+                        {
+                            return result;
+                        }
+
+                        // This transaction has not read this module before, or the entry has not
+                        // yet been verified. For verification, lock the cache.
+                        let mut module_cache =
+                            state.versioned_map.code_cache().module_cache().lock();
+
+                        // The entry may not exist, in which case return early.
+                        let module_id = ModuleId::new(*address, module_name.to_owned());
+                        let entry = match module_cache
+                            .fetch_module_or_store_with(&module_id, || {
+                                self.fetch_versioned_base_module_entry(&module_id)
+                            })? {
+                            Some(entry) => entry,
+                            None => {
+                                state
+                                    .captured_reads
+                                    .borrow_mut()
+                                    .capture_block_cache_read(module_id, None);
+                                module_cache.unlock();
+                                return Ok(None);
+                            },
+                        };
+
+                        // Also return early if the entry has already been verified.
+                        if entry.is_verified() {
+                            // We do not need to capture this read, because the other thread that
+                            // verified this module has put it there.
+                            // TODO(loader_v2): Add a paranoid check here.
+                            module_cache.unlock();
+                            return Ok(Some(entry.verified_module()?.clone()));
+                        }
+
+                        // Otherwise, we need to verify and load the transitive closure.
+
+                        let mut visited = HashSet::new();
+                        visited.insert(module_id.clone());
+
+                        let module = ParallelState::<T, X>::visit_dependencies_and_verify(
+                            entry.clone(),
+                            &module_id,
+                            &mut visited,
+                            self.runtime_environment,
+                            &mut module_cache,
+                            &|id| self.fetch_versioned_base_module_entry(id),
+                        )?;
+
+                        let verified_entry = module_cache.fetch_module_or_store_with(&module_id, || self.fetch_versioned_base_module_entry(&module_id))?.ok_or_else(|| {
```

Comment:
why we need to read from the cache again instead of just using the entry from above?

Reply:
No need, with a single traversal this goes away